agent/dockerclient/dockerapi/docker_client.go

```diff
@@ -838,36 +839,54 @@ func getMetadataHealthCheck(dockerContainer *types.ContainerJSON) apicontainer.H
 
 // Listen to the docker event stream for container changes and pass them up
 func (dg *dockerGoClient) ContainerEvents(ctx context.Context) (<-chan DockerContainerChangeEvent, error) {
-	client, err := dg.dockerClient()
+	client, err := dg.sdkDockerClient()
 	if err != nil {
 		return nil, err
 	}
-	dockerEvents := make(chan *docker.APIEvents, dockerEventBufferSize)
-	events := make(chan *docker.APIEvents)
+
+	events := make(chan *events.Message)
 	buffer := NewInfiniteBuffer()
 
-	err = client.AddEventListener(dockerEvents)
-	if err != nil {
-		seelog.Errorf("DockerGoClient: unable to add a docker event listener: %v", err)
-		return nil, err
-	}
+	derivedCtx, cancel := context.WithCancel(ctx)
+	dockerEvents, eventErr := client.Events(derivedCtx, types.EventsOptions{})
+
+	// Cache the event from docker client. Channel closes when an error is passed to eventErr.
+	go buffer.StartListening(derivedCtx, dockerEvents)
+	// Receive errors from channels. If error thrown is not EOF, log and reopen channel.
 	go func() {
-		<-ctx.Done()
-		client.RemoveEventListener(dockerEvents)
+		for {
+			select {
+			case err := <-eventErr:
+				// If an error is returned, we need to reopen channel to continue listening
+				if err == io.EOF {
```

Comment:
For [fsouza client implementation](https://github.com/fsouza/go-dockerclient/blob/11e1e492616b3d3b7046fceb0f1e7e02f9eef0d8/event.go#L229), when client gets an error, it just returns, should we keep the same behavior?

Reply:
See https://github.com/docker/go-docker/blob/4daae26030ad00e348edddff9767924ae57a3b82/events.go#L16L19
I think in fsouza implementation it's returning from monitoring the events, but it is not closing the event stream itself.

Reply:
For fsouza implementation, inside `monitorEvents`, I think `eventState.C` is the event event stream. And there are three cases to handle the events:
1. Receive an event, it will call `sendEvent` to handle it. 
2. Receive an EOF, it will call `eventState.disableEventMonitoring()` and then `return`, so that it doesn't receive events anymore - this is equivalent to close the event stream.
3. Receive error other than EOF: it will restart the `monitorEvents`.

See [docker events doc](https://docs.docker.com/engine/reference/commandline/events/), there are options like `since` and `until`, when you run the events command with these options, the call to docker dameon is not a blocking call, when you receive the last event, you will end up with an EOF error. But for the default options (Agent case), the `events` API call is a blocking call, which I guess will never end up with an EOF error.

Reply:
Yep you are totally right. Event with no chance of EOF error, I'll duplicate what we were doing with go-dockerclient.